In [1]:
#Imports
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import duckdb as db
import openpyxl
import os
os.makedirs('public/charts', exist_ok=True)
print("All imports successful")

All imports successful


In [2]:

#Load the data
file_path = 'data\WHR26_Data.xlsx'
sheet = 'Data for Figure 2.1 (2011–2025)'

df = pd.read_excel(file_path, sheet_name=sheet)

years_dict = {year: group for year, group in df.groupby('Year')}

print("Data loaded successfully")
print("Shape: ", df.shape)
print("Columns: ", df.columns)
print("First 5 rows: ")
print(df.head())

<>:2: SyntaxWarning: "\W" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\W"? A raw string is also an option.
<>:2: SyntaxWarning: "\W" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\W"? A raw string is also an option.
C:\Users\aliha\AppData\Local\Temp\ipykernel_29468\1129403684.py:2: SyntaxWarning: "\W" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\W"? A raw string is also an option.
  file_path = 'data\WHR26_Data.xlsx'


Data loaded successfully
Shape:  (2116, 13)
Columns:  Index(['Year', 'Rank', 'Country name', 'Life evaluation (3-year average)',
       'Lower whisker', 'Upper whisker', 'Explained by: Log GDP per capita',
       'Explained by: Social support', 'Explained by: Healthy life expectancy',
       'Explained by: Freedom to make life choices',
       'Explained by: Generosity', 'Explained by: Perceptions of corruption',
       'Dystopia + residual'],
      dtype='str')
First 5 rows: 
   Year  Rank Country name  Life evaluation (3-year average)  Lower whisker  \
0  2025     1      Finland                             7.764          7.690   
1  2025     2      Iceland                             7.540          7.449   
2  2025     3      Denmark                             7.539          7.446   
3  2025     4   Costa Rica                             7.439          7.356   
4  2025     5       Sweden                             7.255          7.172   

   Upper whisker  Explained by: Log GDP per

In [3]:
df_2025 = years_dict[2025]
df_2024 = years_dict[2024]
df_2023 = years_dict[2023]
df_2022 = years_dict[2022]
df_2021 = years_dict[2021]
df_2020 = years_dict[2020]
df_2019 = years_dict[2019]
df_2018 = years_dict[2018]
df_2017 = years_dict[2017]
df_2016 = years_dict[2016]
df_2015 = years_dict[2015]

In [4]:
print("Sample data from 2015:")
display(df_2015.head(2))

print("\nSample data from 2019:")
display(df_2019.head(2))

print("\nSample data from 2025:")
display(df_2025.head(2))

Sample data from 2015:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
1489,2015,1,Denmark,7.526,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1490,2015,2,Switzerland,7.509,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN



Sample data from 2019:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
869,2019,1,Finland,7.8087,7.747634,7.869766,1.285190,1.499526,0.961271,0.662317,0.159670,0.477857,2.762835
870,2019,2,Denmark,7.6456,7.579955,7.711245,1.326949,1.503449,0.979333,0.665040,0.242793,0.495260,2.432741



Sample data from 2025:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
0,2025,1,Finland,7.764,7.690,7.837,1.915,1.638,0.939,1.105,0.093,0.491,1.582
1,2025,2,Iceland,7.540,7.449,7.630,1.971,1.720,0.996,1.105,0.187,0.187,1.373


In [5]:
#2015 Dataframe looks empty? Let's check the unique years in the dataset to confirm if other years have this issue.
print("Sample data from 2015:")
display(df_2015.head(2))
print("Sample data from 2016:")
display(df_2016.head(2))
print("Sample data from 2017:")
display(df_2017.head(2))
print("Sample data from 2018:")
display(df_2018.head(2))
print("Sample data from 2019:")
display(df_2019.head(2))
print("Sample data from 2020:")
display(df_2020.head(2))
print("Sample data from 2021:")
display(df_2021.head(2))
print("Sample data from 2022:")
display(df_2022.head(2))
print("Sample data from 2023:")
display(df_2023.head(2))
print("Sample data from 2024:")
display(df_2024.head(2))
print("Sample data from 2025:")
display(df_2025.head(2))

Sample data from 2015:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
1489,2015,1,Denmark,7.526,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1490,2015,2,Switzerland,7.509,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Sample data from 2016:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
1334,2016,1,Norway,7.537,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1335,2016,2,Denmark,7.522,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Sample data from 2017:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
1178,2017,1,Finland,7.632,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1179,2017,2,Norway,7.594,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Sample data from 2018:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
1022,2018,1,Finland,7.769,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1023,2018,2,Denmark,7.600,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Sample data from 2019:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
869,2019,1,Finland,7.8087,7.747634,7.869766,1.285190,1.499526,0.961271,0.662317,0.159670,0.477857,2.762835
870,2019,2,Denmark,7.6456,7.579955,7.711245,1.326949,1.503449,0.979333,0.665040,0.242793,0.495260,2.432741


Sample data from 2020:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
720,2020,1,Finland,7.842,7.780,7.904,1.446,1.106,0.741,0.691,0.124,0.481,3.253
721,2020,2,Denmark,7.620,7.552,7.687,1.502,1.108,0.763,0.686,0.208,0.485,2.868


Sample data from 2021:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
574,2021,1,Finland,7.821,7.756,7.886,1.892,1.258,0.775,0.736,0.109,0.534,2.518
575,2021,2,Denmark,7.636,7.563,7.710,1.953,1.243,0.777,0.719,0.188,0.532,2.226


Sample data from 2022:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
437,2022,1,Finland,7.804,7.733,7.875,1.888,1.585,0.535,0.772,0.126,0.535,2.363
438,2022,2,Denmark,7.586,7.506,7.667,1.949,1.548,0.537,0.734,0.208,0.525,2.084


Sample data from 2023:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
294,2023,1,Finland,7.741,7.667,7.815,1.844,1.572,0.695,0.859,0.142,0.546,2.082
295,2023,2,Denmark,7.583,7.500,7.665,1.908,1.520,0.699,0.823,0.204,0.548,1.881


Sample data from 2024:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
147,2024,1,Finland,7.736,7.662,7.810,1.749,1.783,0.824,0.986,0.11,0.502,1.782
148,2024,2,Denmark,7.521,7.431,7.611,1.825,1.748,0.820,0.955,0.15,0.488,1.535


Sample data from 2025:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
0,2025,1,Finland,7.764,7.690,7.837,1.915,1.638,0.939,1.105,0.093,0.491,1.582
1,2025,2,Iceland,7.540,7.449,7.630,1.971,1.720,0.996,1.105,0.187,0.187,1.373


In [6]:
#Looks like 2015-2018 have no actual data, just the column names
#Gonna have to get everyhting from 2015-18 manually

In [7]:
df_2015 = pd.read_csv(r'data\2015.csv')
df_2016 = pd.read_csv(r'data\2016.csv')
df_2017 = pd.read_csv(r'data\2017.csv')
df_2018 = pd.read_csv(r'data\2018.csv')

In [8]:
print("Sample data from 2015:")
display(df_2015.head(2))
print("Sample data from 2016:")
display(df_2016.head(2))
print("Sample data from 2017:")
display(df_2017.head(2))
print("Sample data from 2018:")
display(df_2018.head(2))
print("Sample data from 2019:")
display(df_2019.head(2))
print("Sample data from 2020:")
display(df_2020.head(2))
print("Sample data from 2021:")
display(df_2021.head(2))
print("Sample data from 2022:")
display(df_2022.head(2))
print("Sample data from 2023:")
display(df_2023.head(2))
print("Sample data from 2024:")
display(df_2024.head(2))
print("Sample data from 2025:")
display(df_2025.head(2))

Sample data from 2015:


,Country,Region,Happiness Rank,Happiness Score,Standard Error,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual
0,Switzerland,Western Europe,1,7.587,0.03411,1.39651,1.34951,0.94143,0.66557,0.41978,0.29678,2.51738
1,Iceland,Western Europe,2,7.561,0.04884,1.30232,1.40223,0.94784,0.62877,0.14145,0.43630,2.70201


Sample data from 2016:


,Country,Region,Happiness Rank,Happiness Score,Lower Confidence Interval,Upper Confidence Interval,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual
0,Denmark,Western Europe,1,7.526,7.460,7.592,1.44178,1.16374,0.79504,0.57941,0.44453,0.36171,2.73939
1,Switzerland,Western Europe,2,7.509,7.428,7.590,1.52733,1.14524,0.86303,0.58557,0.41203,0.28083,2.69463


Sample data from 2017:


,Country,Happiness.Rank,Happiness.Score,Whisker.high,Whisker.low,Economy..GDP.per.Capita.,Family,Health..Life.Expectancy.,Freedom,Generosity,Trust..Government.Corruption.,Dystopia.Residual
0,Norway,1,7.537,7.594445,7.479556,1.616463,1.533524,0.796667,0.635423,0.362012,0.315964,2.277027
1,Denmark,2,7.522,7.581728,7.462272,1.482383,1.551122,0.792566,0.626007,0.355280,0.400770,2.313707


Sample data from 2018:


,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
0,1,Finland,7.632,1.305,1.592,0.874,0.681,0.202,0.393
1,2,Norway,7.594,1.456,1.582,0.861,0.686,0.286,0.340


Sample data from 2019:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
869,2019,1,Finland,7.8087,7.747634,7.869766,1.285190,1.499526,0.961271,0.662317,0.159670,0.477857,2.762835
870,2019,2,Denmark,7.6456,7.579955,7.711245,1.326949,1.503449,0.979333,0.665040,0.242793,0.495260,2.432741


Sample data from 2020:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
720,2020,1,Finland,7.842,7.780,7.904,1.446,1.106,0.741,0.691,0.124,0.481,3.253
721,2020,2,Denmark,7.620,7.552,7.687,1.502,1.108,0.763,0.686,0.208,0.485,2.868


Sample data from 2021:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
574,2021,1,Finland,7.821,7.756,7.886,1.892,1.258,0.775,0.736,0.109,0.534,2.518
575,2021,2,Denmark,7.636,7.563,7.710,1.953,1.243,0.777,0.719,0.188,0.532,2.226


Sample data from 2022:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
437,2022,1,Finland,7.804,7.733,7.875,1.888,1.585,0.535,0.772,0.126,0.535,2.363
438,2022,2,Denmark,7.586,7.506,7.667,1.949,1.548,0.537,0.734,0.208,0.525,2.084


Sample data from 2023:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
294,2023,1,Finland,7.741,7.667,7.815,1.844,1.572,0.695,0.859,0.142,0.546,2.082
295,2023,2,Denmark,7.583,7.500,7.665,1.908,1.520,0.699,0.823,0.204,0.548,1.881


Sample data from 2024:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
147,2024,1,Finland,7.736,7.662,7.810,1.749,1.783,0.824,0.986,0.11,0.502,1.782
148,2024,2,Denmark,7.521,7.431,7.611,1.825,1.748,0.820,0.955,0.15,0.488,1.535


Sample data from 2025:


,Year,Rank,Country name,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
0,2025,1,Finland,7.764,7.690,7.837,1.915,1.638,0.939,1.105,0.093,0.491,1.582
1,2025,2,Iceland,7.540,7.449,7.630,1.971,1.720,0.996,1.105,0.187,0.187,1.373


In [9]:
print("2015 Statistical Summary:")
display(df_2015.describe())

print("\n2020 Statistical Summary:")
display(df_2020.describe())

print("\n2025 Statistical Summary:")
display(df_2025.describe())

2015 Statistical Summary:


,Happiness Rank,Happiness Score,Standard Error,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual
count,158.000000,158.000000,158.000000,158.000000,158.000000,158.000000,158.000000,158.000000,158.000000,158.000000
mean,79.493671,5.375734,0.047885,0.846137,0.991046,0.630259,0.428615,0.143422,0.237296,2.098977
std,45.754363,1.145010,0.017146,0.403121,0.272369,0.247078,0.150693,0.120034,0.126685,0.553550
min,1.000000,2.839000,0.018480,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.328580
25%,40.250000,4.526000,0.037268,0.545808,0.856823,0.439185,0.328330,0.061675,0.150553,1.759410
50%,79.500000,5.232500,0.043940,0.910245,1.029510,0.696705,0.435515,0.107220,0.216130,2.095415
75%,118.750000,6.243750,0.052300,1.158448,1.214405,0.811013,0.549092,0.180255,0.309883,2.462415
max,158.000000,7.587000,0.136930,1.690420,1.402230,1.025250,0.669730,0.551910,0.795880,3.602140



2020 Statistical Summary:


,Year,Rank,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
count,149.0,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000,149.000000
mean,2020.0,75.000000,5.532839,5.417631,5.648007,0.977161,0.793315,0.520161,0.498711,0.178047,0.135141,2.430329
std,0.0,43.156691,1.073924,1.094879,1.054330,0.404740,0.258871,0.213019,0.137888,0.098270,0.114361,0.537645
min,2020.0,1.000000,2.523000,2.449000,2.596000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.648000
25%,2020.0,38.000000,4.852000,4.706000,4.991000,0.666000,0.647000,0.357000,0.409000,0.105000,0.060000,2.138000
50%,2020.0,75.000000,5.534000,5.413000,5.625000,1.025000,0.832000,0.571000,0.514000,0.164000,0.101000,2.509000
75%,2020.0,112.000000,6.255000,6.128000,6.344000,1.323000,0.996000,0.665000,0.603000,0.239000,0.174000,2.794000
max,2020.0,149.000000,7.842000,7.780000,7.904000,1.751000,1.172000,0.897000,0.716000,0.541000,0.547000,3.482000



2025 Statistical Summary:


,Year,Rank,Life evaluation (3-year average),Lower whisker,Upper whisker,Explained by: Log GDP per capita,Explained by: Social support,Explained by: Healthy life expectancy,Explained by: Freedom to make life choices,Explained by: Generosity,Explained by: Perceptions of corruption,Dystopia + residual
count,147.0,147.000000,147.000000,147.000000,147.000000,147.000000,147.000000,146.000000,146.000000,147.000000,147.000000,145.000000
mean,2025.0,74.000000,5.649857,5.539401,5.760259,1.534932,1.204762,0.616123,0.882308,0.106306,0.148204,1.163200
std,0.0,42.579338,1.138310,1.150668,1.127029,0.365199,0.355028,0.260254,0.181011,0.052300,0.109909,0.520333
min,2025.0,1.000000,1.446000,1.348000,1.544000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,-0.295000
25%,2025.0,37.500000,4.666000,4.546000,4.784000,1.290000,0.996000,0.443750,0.801750,0.062500,0.073000,0.913000
50%,2025.0,74.000000,5.937000,5.829000,6.064000,1.565000,1.292000,0.607000,0.913500,0.106000,0.115000,1.164000
75%,2025.0,110.500000,6.576000,6.460000,6.675000,1.832000,1.481000,0.827750,1.014750,0.143500,0.188000,1.496000
max,2025.0,147.000000,7.764000,7.690000,7.837000,2.167000,1.720000,1.238000,1.147000,0.295000,0.512000,2.701000


In [10]:
#Observations:
#Number of countries went down?
#Average happiness went down
#Column names changed -- this will be annoying


In [11]:
# ── 2015–2018: loaded from individual CSVs ───────────────────────────────────

# For 2015
df_2015_std = df_2015.copy()
df_2015_std['Year'] = 2015
df_2015_std = df_2015_std.rename(columns={
    'Country'                      : 'Country',
    'Happiness Rank'               : 'Happiness_Rank',
    'Happiness Score'              : 'Happiness_Score',
    'Economy (GDP per Capita)'     : 'GDP_per_Capita',
    'Family'                       : 'Social_Support',
    'Health (Life Expectancy)'     : 'Life_Expectancy',
    'Freedom'                      : 'Freedom',
    'Trust (Government Corruption)': 'Corruption',
    'Generosity'                   : 'Generosity',
    'Dystopia Residual'            : 'Dystopia_Residual',
})

# For 2016
df_2016_std = df_2016.copy()
df_2016_std['Year'] = 2016
df_2016_std = df_2016_std.rename(columns={
    'Country'                      : 'Country',
    'Happiness Rank'               : 'Happiness_Rank',
    'Happiness Score'              : 'Happiness_Score',
    'Economy (GDP per Capita)'     : 'GDP_per_Capita',
    'Family'                       : 'Social_Support',
    'Health (Life Expectancy)'     : 'Life_Expectancy',
    'Freedom'                      : 'Freedom',
    'Trust (Government Corruption)': 'Corruption',
    'Generosity'                   : 'Generosity',
    'Dystopia Residual'            : 'Dystopia_Residual',
})

# For 2017
df_2017_std = df_2017.copy()
df_2017_std['Year'] = 2017
df_2017_std = df_2017_std.rename(columns={
    'Country'                      : 'Country',
    'Happiness.Rank'               : 'Happiness_Rank',
    'Happiness.Score'              : 'Happiness_Score',
    'Economy..GDP.per.Capita.'     : 'GDP_per_Capita',
    'Family'                       : 'Social_Support',
    'Health..Life.Expectancy.'     : 'Life_Expectancy',
    'Freedom'                      : 'Freedom',
    'Trust..Government.Corruption.': 'Corruption',
    'Generosity'                   : 'Generosity',
    'Dystopia.Residual'            : 'Dystopia_Residual',
})

# For 2018
df_2018_std = df_2018.copy()
df_2018_std['Year'] = 2018
df_2018_std = df_2018_std.rename(columns={
    'Country or region'            : 'Country',
    'Overall rank'                 : 'Happiness_Rank',
    'Score'                        : 'Happiness_Score',
    'GDP per capita'               : 'GDP_per_Capita',
    'Social support'               : 'Social_Support',
    'Healthy life expectancy'      : 'Life_Expectancy',
    'Freedom to make life choices' : 'Freedom',
    'Perceptions of corruption'    : 'Corruption',
    'Generosity'                   : 'Generosity',
})

# ── 2019–2025: all from WHR26_Data.xlsx (one sheet, filter by Year) ──────────
# Actual column names in that sheet are different from the CSVs above.

WHR26_RENAME = {
    'Country name'                              : 'Country',
    'Rank'                                      : 'Happiness_Rank',
    'Life evaluation (3-year average)'          : 'Happiness_Score',
    'Explained by: Log GDP per capita'          : 'GDP_per_Capita',
    'Explained by: Social support'              : 'Social_Support',
    'Explained by: Healthy life expectancy'     : 'Life_Expectancy',
    'Explained by: Freedom to make life choices': 'Freedom',
    'Explained by: Perceptions of corruption'   : 'Corruption',
    'Explained by: Generosity'                  : 'Generosity',
    'Dystopia + residual'                       : 'Dystopia_Residual',
    'Lower whisker'                             : 'Lower_Whisker',
    'Upper whisker'                             : 'Upper_Whisker',
}

df_2019_std = df[df['Year'] == 2019].copy().rename(columns=WHR26_RENAME)
df_2020_std = df[df['Year'] == 2020].copy().rename(columns=WHR26_RENAME)
df_2021_std = df[df['Year'] == 2021].copy().rename(columns=WHR26_RENAME)
df_2022_std = df[df['Year'] == 2022].copy().rename(columns=WHR26_RENAME)
df_2023_std = df[df['Year'] == 2023].copy().rename(columns=WHR26_RENAME)
df_2024_std = df[df['Year'] == 2024].copy().rename(columns=WHR26_RENAME)
df_2025_std = df[df['Year'] == 2025].copy().rename(columns=WHR26_RENAME)

In [12]:
#Missing values
all_dfs = [df_2015_std, df_2016_std, df_2017_std, df_2018_std, df_2019_std, df_2020_std, df_2021_std, df_2022_std, df_2023_std, df_2024_std, df_2025_std]
#Check for missing values in each dataframe
for i, df in enumerate(all_dfs):
    missing = all_dfs[i].isnull().sum()
    year = 2015 + i

    if missing.sum() > 0:
        print(f"Missing values in {year} dataset:")
        print(missing[missing > 0])
    else:
        print(f"No missing values in {year} dataset.")

No missing values in 2015 dataset.
No missing values in 2016 dataset.
No missing values in 2017 dataset.
Missing values in 2018 dataset:
Corruption    1
dtype: int64
No missing values in 2019 dataset.
No missing values in 2020 dataset.
No missing values in 2021 dataset.
Missing values in 2022 dataset:
Life_Expectancy      1
Dystopia_Residual    1
dtype: int64
Missing values in 2023 dataset:
GDP_per_Capita       3
Social_Support       3
Life_Expectancy      3
Freedom              3
Generosity           3
Corruption           3
Dystopia_Residual    3
dtype: int64
Missing values in 2024 dataset:
Life_Expectancy      1
Freedom              1
Corruption           1
Dystopia_Residual    3
dtype: int64
Missing values in 2025 dataset:
Life_Expectancy      1
Freedom              1
Dystopia_Residual    2
dtype: int64


In [13]:
#we will fill missing values with the median of each column for that year, since median is more robust to outliers than mean.
for i in range(len(all_dfs)):
    year = 2015 + i
    for col in all_dfs[i].columns:
        if all_dfs[i][col].isnull().sum() > 0:
            median_value = all_dfs[i][col].median()
            all_dfs[i][col] = all_dfs[i][col].fillna(median_value)
            print(f"Filled missing values in column '{col}' of {year} dataset with median value {median_value}.")
            
#print the number of missing values after filling
for i, df in enumerate(all_dfs):
    missing = all_dfs[i].isnull().sum()
    year = 2015 + i

    if missing.sum() > 0:
        print(f"Missing values in {year} dataset:")
        print(missing[missing > 0])
    else:
        print(f"No missing values in {year} dataset.")

Filled missing values in column 'Corruption' of 2018 dataset with median value 0.082.
Filled missing values in column 'Life_Expectancy' of 2022 dataset with median value 0.3895.
Filled missing values in column 'Dystopia_Residual' of 2022 dataset with median value 1.8485.
Filled missing values in column 'GDP_per_Capita' of 2023 dataset with median value 1.4315.
Filled missing values in column 'Social_Support' of 2023 dataset with median value 1.2375.
Filled missing values in column 'Life_Expectancy' of 2023 dataset with median value 0.5495000000000001.
Filled missing values in column 'Freedom' of 2023 dataset with median value 0.641.
Filled missing values in column 'Generosity' of 2023 dataset with median value 0.1365.
Filled missing values in column 'Corruption' of 2023 dataset with median value 0.1205.
Filled missing values in column 'Dystopia_Residual' of 2023 dataset with median value 1.6444999999999999.
Filled missing values in column 'Life_Expectancy' of 2024 dataset with median v

In [14]:
#Merge all dataframes into one master dataframe for easier analysis
master_df = pd.concat(all_dfs, ignore_index=True)
print("Master dataframe created successfully with shape: ", master_df.shape)
print("Columns in master dataframe: ", master_df.columns)
print("First 5 rows of master dataframe: ")
print(master_df.head())

Master dataframe created successfully with shape:  (1648, 19)
Columns in master dataframe:  Index(['Country', 'Region', 'Happiness_Rank', 'Happiness_Score',
       'Standard Error', 'GDP_per_Capita', 'Social_Support', 'Life_Expectancy',
       'Freedom', 'Corruption', 'Generosity', 'Dystopia_Residual', 'Year',
       'Lower Confidence Interval', 'Upper Confidence Interval',
       'Whisker.high', 'Whisker.low', 'Lower_Whisker', 'Upper_Whisker'],
      dtype='str')
First 5 rows of master dataframe: 
       Country          Region  Happiness_Rank  Happiness_Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  GDP_per_Capita  Social_Support  Life_Expectancy  Freedom  \
0         0.03

In [15]:
total_unique_countries = master_df['Country'].nunique()
print(f"Total unique countries across all years: {total_unique_countries}")

country_name_mapping = {
    'United States': 'United States of America',
    'Congo (Kinshasa)': 'Democratic Republic of the Congo',
    'Congo (Brazzaville)': 'Republic of Congo',
    'North Cyprus': 'Northern Cyprus',
    'Hong Kong S.A.R. of China': 'Hong Kong',
    'Hong Kong S.A.R., China': 'Hong Kong',
    'Taiwan Province of China': 'Taiwan',
    'Palestinian Territories': 'Palestine'
}

# Apply the mapping to standardize country names
master_df['Country'] = master_df['Country'].replace(country_name_mapping)

# Check unique countries after harmonization
harmonized_unique_countries = master_df['Country'].nunique()
print(f"Unique countries after harmonization: {harmonized_unique_countries}")

Total unique countries across all years: 184
Unique countries after harmonization: 181


---
## Exploratory Data Analysis

With a clean, unified dataset spanning **2015–2025** across up to 158 countries, we now explore the key questions:
- How has global happiness trended over time?
- Which countries consistently lead — and which lag?
- What factors actually drive happiness scores?
- How did COVID reshape the global wellbeing landscape?


In [16]:
# ── Portfolio dark theme (matches website palette) ────────────────────────────
DARK = dict(
    paper_bgcolor='#111111',
    plot_bgcolor='#111111',
    font=dict(color='#f0ece4', family='IBM Plex Mono'),
)
AXIS = dict(gridcolor='#1e1e1e', linecolor='#2a2a2a', zerolinecolor='#1e1e1e')
ACCENT    = '#c8a96e'
ACCENT2   = '#8ba8b5'
PALETTE   = ['#c8a96e','#8ba8b5','#a0856b','#6b8a6b','#b57a7a','#7a7ab5','#b5a07a']


### 1. Global Happiness Trends (2019–2025)

We restrict this trend analysis to 2019–2025 to ensure methodological consistency — the WHR26 dataset provides clean, comparable life evaluation scores from 2019 onward, while 2015–2018 data sourced from annual CSVs uses a different scoring methodology.

Despite the turbulence of a global pandemic, the global average happiness score has proven **surprisingly resilient**, rising from ~5.47 in 2019 to ~5.65 in 2025.

In [17]:
# Visualization 1 — Global happiness trends (2019–2025)
import plotly.express as px
import plotly.graph_objects as go
df_trend = master_df[master_df['Year'] >= 2019].groupby('Year')['Happiness_Score'].mean().reset_index()

fig = px.line(
    df_trend, x='Year', y='Happiness_Score',
    title='Global Average Happiness Score (2019–2025)',
    markers=True,
)
fig.update_traces(line=dict(color=ACCENT, width=2.5), marker=dict(color=ACCENT, size=7))
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Avg. Happiness Score',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1))
fig.write_html('public/charts/whr_global_trend.html', full_html=True, include_plotlyjs='cdn')
fig.show()


### 2. Top 10 & Bottom 10 Happiest Countries (2025)

The 2025 rankings reinforce a pattern that has held for over a decade: **Nordic countries dominate the top**, while Sub-Saharan African and conflict-affected nations cluster at the bottom. Finland claims the #1 position for the eighth consecutive year.

The gap between the top and bottom is striking — a difference of roughly 5 points on a 10-point scale, representing a fundamentally different lived experience.

In [18]:
# Visualization 2 — Top 10 & Bottom 10 happiest countries (2025)
df_2025_sorted = df_2025_std.sort_values('Happiness_Score', ascending=False)
top_10    = df_2025_sorted.head(10)
bottom_10 = df_2025_sorted.tail(10)

fig_top = px.bar(
    top_10, x='Country', y='Happiness_Score',
    title='Top 10 Happiest Countries — 2025',
    text=top_10['Happiness_Score'].round(3),
)
fig_top.update_traces(marker_color=ACCENT, textposition='outside', textfont=dict(color='#f0ece4'))
fig_top.update_layout(**DARK, xaxis_title='', yaxis_title='Happiness Score')
fig_top.write_html('public/charts/whr_top_10.html', full_html=True, include_plotlyjs='cdn')
fig_top.show()

fig_bottom = px.bar(
    bottom_10.sort_values('Happiness_Score'), x='Country', y='Happiness_Score',
    title='Bottom 10 Happiest Countries — 2025',
    text=bottom_10['Happiness_Score'].round(3),
)
fig_bottom.update_traces(marker_color='#b57a7a', textposition='outside', textfont=dict(color='#f0ece4'))
fig_bottom.update_layout(**DARK, xaxis_title='', yaxis_title='Happiness Score')
fig_bottom.show()


### 3. Evolution of the Top 5 Countries (2015–2025)

Tracking the top 5 countries from 2025 backward reveals both **stability and subtle movement**. Finland, Denmark, and Iceland have traded positions but remained in elite company throughout the decade. The consistency suggests that once a country achieves high wellbeing, the structural factors maintaining it are deeply embedded.

In [19]:
# Visualization 3 — Evolution of top 5 countries over time
top_5_countries = df_2025_std.sort_values('Happiness_Score', ascending=False).head(5)['Country'].tolist()
print("Top 5 in 2025:", top_5_countries)

plot_df = master_df[master_df['Country'].isin(top_5_countries)]
fig = px.line(
    plot_df, x='Year', y='Happiness_Score', color='Country',
    title='Top 5 Happiest Countries — Score Evolution (2015–2025)',
    markers=True, color_discrete_sequence=PALETTE,
)
fig.update_traces(line=dict(width=2))
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Happiness Score',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1),
                  legend=dict(bgcolor='#111111', bordercolor='#1e1e1e'))
fig.show()


Top 5 in 2025: ['Finland', 'Iceland', 'Denmark', 'Costa Rica', 'Sweden']


### 4. Correlation Matrix — Happiness & Contributing Factors

Across all years and countries, three factors stand out as the strongest correlates of happiness: **GDP per capita, social support, and life expectancy**. Corruption shows the strongest *negative* correlation — countries where citizens distrust institutions score markedly lower.

Generosity and freedom show weaker but still positive relationships, suggesting they matter more at the margin once basic material and social needs are met.

In [20]:
# Visualization 4 — Correlation matrix: happiness vs contributing factors
corr_cols = ['Happiness_Score','GDP_per_Capita','Social_Support','Life_Expectancy','Freedom','Corruption','Generosity']
correlation_matrix = master_df[corr_cols].corr().round(2)

fig = px.imshow(
    correlation_matrix,
    color_continuous_scale=[[0,'#3a2a1a'],[0.5,'#111111'],[1,'#c8a96e']],
    title='Correlation Matrix — Happiness & Contributing Factors (2015–2025)',
    text_auto=True,
    zmin=-1, zmax=1,
)
fig.update_layout(**DARK, xaxis_title='', yaxis_title='',
                  coloraxis_colorbar=dict(tickfont=dict(color='#7a7570')))
fig.show()


### 5. How Correlations Evolved Over Time

The relationship between happiness and its drivers is not static. Post-COVID, **GDP per capita and social support gained explanatory power** — consistent with the hypothesis that economic disruption and social isolation made financial security and community more salient to people's sense of wellbeing.

Life expectancy's correlation weakened slightly in the 2020–2022 window, which may reflect the unique nature of COVID-era mortality — concentrated in older populations in ways that didn't uniformly suppress national happiness scores.

In [21]:
# Visualization 5 — How correlations with happiness evolved year-over-year
correlation_over_time = []
for year in sorted(master_df['Year'].unique()):
    yd = master_df[master_df['Year'] == year]
    correlation_over_time.append({
        'Year'           : year,
        'GDP_per_Capita' : yd['Happiness_Score'].corr(yd['GDP_per_Capita']),
        'Social_Support' : yd['Happiness_Score'].corr(yd['Social_Support']),
        'Life_Expectancy': yd['Happiness_Score'].corr(yd['Life_Expectancy']),
        'Freedom'        : yd['Happiness_Score'].corr(yd['Freedom']),
        'Corruption'     : yd['Happiness_Score'].corr(yd['Corruption']),
        'Generosity'     : yd['Happiness_Score'].corr(yd['Generosity']),
    })

correlation_over_time_df = pd.DataFrame(correlation_over_time)

fig = px.line(
    correlation_over_time_df, x='Year',
    y=['GDP_per_Capita','Social_Support','Life_Expectancy','Freedom','Corruption','Generosity'],
    title='Correlation with Happiness Score — Year-over-Year (2015–2025)',
    markers=True, color_discrete_sequence=PALETTE,
)
fig.update_traces(line=dict(width=2))
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Correlation Coefficient',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1),
                  legend=dict(bgcolor='#111111', bordercolor='#1e1e1e'))
fig.show()


### 6. Distribution of Happiness Scores Across Years

The violin plot reveals the **shape** of global happiness — not just the average. The distribution has remained relatively stable, but with a notable narrowing of the lower tail from 2022 onward, suggesting that the worst-performing countries have seen modest improvements even as top performers held steady.

In [22]:
# Visualization 6 — Distribution of happiness scores across the years
fig = px.violin(
    master_df, x='Year', y='Happiness_Score',
    title='Distribution of Happiness Scores Across the Years (2015–2025)',
    box=True, points=False, color_discrete_sequence=[ACCENT],
)
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Happiness Score',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1))
fig.show()


### 7. G7 Countries — Trends & Global Standing

The G7 represents seven of the world's largest and wealthiest economies. Yet their happiness scores tell a more complicated story: the **G7 average ranks roughly 10–15th globally**, consistently trailing Nordic countries that are smaller but score higher on social support, freedom, and institutional trust.

This is arguably the most counterintuitive finding in the dataset — economic supremacy does not translate to wellbeing supremacy.

In [23]:
# Visualization 7 — G7 trends and comparison to top 5
g7_countries = ['United States of America','Canada','United Kingdom','Germany','France','Italy','Japan']
g7_df = master_df[master_df['Country'].isin(g7_countries)]

fig = px.line(
    g7_df, x='Year', y='Happiness_Score', color='Country',
    title='G7 Happiness Score Trends (2015–2025)',
    markers=True, color_discrete_sequence=PALETTE,
)
fig.update_traces(line=dict(width=2))
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Happiness Score',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1),
                  legend=dict(bgcolor='#111111', bordercolor='#1e1e1e'))
fig.show()

# G7 average vs top 5
g7_avg = g7_df.groupby('Year')['Happiness_Score'].mean().reset_index()
g7_avg['Country'] = 'G7 Average'
top_5_countries = df_2025_std.sort_values('Happiness_Score', ascending=False).head(5)['Country'].tolist()
plot_df = pd.concat([g7_avg, master_df[master_df['Country'].isin(top_5_countries)]], ignore_index=True)

fig2 = px.line(
    plot_df, x='Year', y='Happiness_Score', color='Country',
    title='G7 Average vs Top 5 Happiest Countries (2015–2025)',
    markers=True, color_discrete_sequence=[ACCENT] + PALETTE[1:],
)
fig2.update_traces(line=dict(width=2))
fig2.update_layout(**DARK, xaxis_title='Year', yaxis_title='Happiness Score',
                   xaxis=dict(**AXIS, tickmode='linear', dtick=1),
                   legend=dict(bgcolor='#111111', bordercolor='#1e1e1e'))
fig2.show()

# Where would G7 rank each year?
for year in sorted(master_df['Year'].unique()):
    g7_score = g7_avg[g7_avg['Year'] == year]['Happiness_Score'].values[0]
    rank = (master_df[(master_df['Year'] == year) & (master_df['Happiness_Score'] > g7_score)].shape[0]) + 1
    print(f"{year}  G7 avg: {g7_score:.3f}  →  would rank #{rank}")


2015  G7 avg: 6.668  →  would rank #28
2016  G7 avg: 6.658  →  would rank #26
2017  G7 avg: 6.614  →  would rank #23
2018  G7 avg: 6.682  →  would rank #22
2019  G7 avg: 6.762  →  would rank #23
2020  G7 avg: 6.769  →  would rank #21
2021  G7 avg: 6.739  →  would rank #20
2022  G7 avg: 6.677  →  would rank #21
2023  G7 avg: 6.584  →  would rank #29
2024  G7 avg: 6.595  →  would rank #33
2025  G7 avg: 6.632  →  would rank #34


### 8. Pre-COVID vs Post-COVID Happiness

The expectation going into this analysis was clear: COVID should have crushed global happiness scores. The data says otherwise. **Post-COVID happiness is actually higher on average** than pre-COVID, which demands an explanation.

The most likely drivers: adaptation effects, government support programs buffering economic harm, and the possibility that the WHR's 3-year rolling average methodology smoothed out the 2020 shock. The pandemic did not permanently break happiness — it may have briefly clarified what people value most.

In [24]:
# Visualization 8 — Pre-COVID vs Post-COVID global comparison
pre_covid  = master_df[master_df['Year'] < 2020]
post_covid = master_df[master_df['Year'] >= 2020]

pre_covid_mean  = pre_covid['Happiness_Score'].mean()
post_covid_mean = post_covid['Happiness_Score'].mean()
print(f"Pre-COVID  avg (2015–2019): {pre_covid_mean:.3f}")
print(f"Post-COVID avg (2020–2025): {post_covid_mean:.3f}")

covid_df = pd.DataFrame({
    'Period'                : ['Pre-COVID (2015–2019)', 'Post-COVID (2020–2025)'],
    'Average_Happiness_Score': [pre_covid_mean, post_covid_mean],
})
fig = px.bar(
    covid_df, x='Period', y='Average_Happiness_Score',
    title='Global Average Happiness — Pre vs Post COVID',
    text=covid_df['Average_Happiness_Score'].round(3),
    color='Period', color_discrete_sequence=[ACCENT2, ACCENT],
)
fig.update_traces(textposition='outside', textfont=dict(color='#f0ece4'))
fig.update_layout(**DARK, xaxis_title='', yaxis_title='Average Happiness Score',
                  showlegend=False)
fig.show()


Pre-COVID  avg (2015–2019): 5.392
Post-COVID avg (2020–2025): 5.564


### 9. Biggest Winners & Losers Since COVID

At the country level, the aggregate stability masks enormous variation. **Afghanistan and Lebanon** saw catastrophic declines — each driven by distinct crises (Taliban takeover and economic collapse respectively) that overwhelmed any pandemic-era resilience.

The biggest gainers are concentrated in Eastern Europe and parts of Africa, reflecting a mix of baseline effects (room to grow from low starting points) and genuine institutional improvements in some cases.

In [25]:
# Visualization 9 — Countries with biggest COVID-era happiness swings
pre_covid_avg  = pre_covid.groupby('Country')['Happiness_Score'].mean().reset_index()
post_covid_avg = post_covid.groupby('Country')['Happiness_Score'].mean().reset_index()
pre_post_df    = pd.merge(pre_covid_avg, post_covid_avg, on='Country',
                          suffixes=('_Pre_COVID','_Post_COVID'))
pre_post_df['Score_Change'] = pre_post_df['Happiness_Score_Post_COVID'] - pre_post_df['Happiness_Score_Pre_COVID']

top_10_increase = pre_post_df.sort_values('Score_Change', ascending=False).head(10)
top_10_decrease = pre_post_df.sort_values('Score_Change').head(10)

print("Biggest increases:")
print(top_10_increase[['Country','Score_Change']].to_string(index=False))
print("\nBiggest decreases:")
print(top_10_decrease[['Country','Score_Change']].to_string(index=False))

# Combined chart
top_10_increase['Direction'] = 'Increase'
top_10_decrease['Direction'] = 'Decrease'
swing_df = pd.concat([top_10_increase, top_10_decrease])

fig = px.bar(
    swing_df.sort_values('Score_Change'), x='Score_Change', y='Country',
    color='Direction', orientation='h',
    title='Biggest Happiness Swings — Pre vs Post COVID',
    color_discrete_map={'Increase': ACCENT, 'Decrease': '#b57a7a'},
)
fig.update_layout(**DARK, xaxis_title='Score Change', yaxis_title='',
                  legend=dict(bgcolor='#111111', bordercolor='#1e1e1e'))
fig.show()


Biggest increases:
 Country  Score_Change
 Armenia      1.009407
  Guinea      0.981340
  Serbia      0.977093
  Kosovo      0.902927
Bulgaria      0.833800
 Georgia      0.809647
 Romania      0.772360
  Belize      0.755000
Slovenia      0.753587
Honduras      0.739227

Biggest decreases:
     Country  Score_Change
     Lebanon     -1.806333
 Afghanistan     -1.499413
Sierra Leone     -1.127180
    DR Congo     -0.878000
      Jordan     -0.851580
     Somalia     -0.763500
    Zimbabwe     -0.696007
        Oman     -0.656000
    Pakistan     -0.635727
      Zambia     -0.625713


---
## Predictive Modelling

Can we predict a country's happiness score from its structural factors? We train three regression models on data from **2015–2022**, evaluate them on **2023–2024** (held-out test set), and then validate the best performer against **2025 actuals**.

The temporal split is deliberate — randomly shuffling years would leak future information into training, producing artificially inflated performance metrics.

**Models:** Linear Regression · Random Forest · XGBoost  
**Baseline:** per-country mean score (last known year)  
**Metrics:** MSE + R²

In [26]:
# Predictive Modelling — train on 2015–2022, evaluate on 2023–2024, predict 2025
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.cluster import KMeans
from sklearn.metrics import mean_squared_error, silhouette_score, r2_score

features   = ['GDP_per_Capita','Social_Support','Life_Expectancy','Freedom','Corruption','Generosity']
train_data = master_df[master_df['Year'] < 2023]
test_data  = master_df[(master_df['Year'] >= 2023) & (master_df['Year'] < 2025)]
X_train, y_train = train_data[features], train_data['Happiness_Score']
X_test,  y_test  = test_data[features],  test_data['Happiness_Score']

# Linear Regression
lr_model        = LinearRegression()
lr_model.fit(X_train, y_train)
lr_predictions  = lr_model.predict(X_test)
lr_mse          = mean_squared_error(y_test, lr_predictions)
print(f"Linear Regression  — MSE: {lr_mse:.3f}, R²: {r2_score(y_test, lr_predictions):.3f}")

# Random Forest
rf_model        = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
rf_predictions  = rf_model.predict(X_test)
rf_mse          = mean_squared_error(y_test, rf_predictions)
print(f"Random Forest      — MSE: {rf_mse:.3f}, R²: {r2_score(y_test, rf_predictions):.3f}")

# XGBoost
xgb_model       = XGBRegressor(n_estimators=100, random_state=42)
xgb_model.fit(X_train, y_train)
xgb_predictions = xgb_model.predict(X_test)
xgb_mse         = mean_squared_error(y_test, xgb_predictions)
print(f"XGBoost            — MSE: {xgb_mse:.3f}, R²: {r2_score(y_test, xgb_predictions):.3f}")

# KMeans
kmeans_model    = KMeans(n_clusters=3, random_state=42)
kmeans_model.fit(X_train)
kmeans_silhouette = silhouette_score(X_train, kmeans_model.labels_)
print(f"KMeans Silhouette  — Score: {kmeans_silhouette:.3f}")

# Best model — validated on 2025
best_model = min([(lr_model, lr_mse),(rf_model, rf_mse),(xgb_model, xgb_mse)], key=lambda x: x[1])[0]
best_model_name = "Linear Regression" if best_model == lr_model else "Random Forest" if best_model == rf_model else "XGBoost"

X_2025          = master_df[master_df['Year'] == 2025][features]
y_2025_actual   = master_df[master_df['Year'] == 2025]['Happiness_Score']
y_2025_predicted = best_model.predict(X_2025)
prediction_mse  = mean_squared_error(y_2025_actual, y_2025_predicted)

print(f"\nBest model: {best_model_name}")
print(f"Validation MSE on 2025 actuals: {prediction_mse:.3f}, R²: {r2_score(y_2025_actual, y_2025_predicted):.3f}")

comparison_df = pd.DataFrame({
    'Country'                  : master_df[master_df['Year'] == 2025]['Country'].values,
    'Actual_Happiness_Score'   : y_2025_actual.values,
    'Predicted_Happiness_Score': y_2025_predicted,
})
print("\nActual vs Predicted — Top 10:")
print(comparison_df.sort_values('Actual_Happiness_Score', ascending=False).head(10).to_string(index=False))


Linear Regression  — MSE: 0.612, R²: 0.546
Random Forest      — MSE: 0.375, R²: 0.722
XGBoost            — MSE: 0.389, R²: 0.711
KMeans Silhouette  — Score: 0.273

Best model: Random Forest
Validation MSE on 2025 actuals: 0.489, R²: 0.620

Actual vs Predicted — Top 10:
    Country  Actual_Happiness_Score  Predicted_Happiness_Score
    Finland                   7.764                   7.354502
    Iceland                   7.540                   6.915718
    Denmark                   7.539                   7.507456
 Costa Rica                   7.439                   6.401945
     Sweden                   7.255                   7.398424
     Norway                   7.242                   7.475512
Netherlands                   7.223                   7.365718
     Israel                   7.187                   6.851156
 Luxembourg                   7.063                   7.309225
Switzerland                   7.018                   7.347016


### Residual Analysis

A well-behaved model shows residuals scattered randomly around zero with no systematic pattern. Funneling (residuals growing with predicted value) would indicate heteroscedasticity — the model struggling more at extreme scores.

In [27]:
# Residual plot — how well does the best model fit?
import plotly.graph_objects as go

y_pred_test = best_model.predict(X_test)
residuals   = y_test.values - y_pred_test

fig = go.Figure()
fig.add_trace(go.Scatter(
    x=y_pred_test, y=residuals, mode='markers',
    marker=dict(color=ACCENT, opacity=0.6, size=6),
    name='Residuals',
))
fig.add_hline(y=0, line=dict(color='#3a3632', width=1, dash='dash'))
fig.update_layout(**DARK,
    title=f'Residual Plot — {best_model_name}',
    xaxis_title='Predicted Score', yaxis_title='Residual (Actual − Predicted)',
)
fig.show()


### Feature Importance

Which factors does the model lean on most? Feature importance from tree-based models reflects **predictive contribution** — how much each variable reduces prediction error when included. This differs from correlation, which only measures linear association.

In [28]:
# Feature importance — what drives happiness scores?
if best_model == rf_model:
    feature_importances = rf_model.feature_importances_
elif best_model == xgb_model:
    feature_importances = xgb_model.feature_importances_
else:
    feature_importances = abs(lr_model.coef_) / abs(lr_model.coef_).sum()

fi_df = pd.DataFrame({'Feature': features, 'Importance': feature_importances}).sort_values('Importance', ascending=True)

fig = px.bar(
    fi_df, x='Importance', y='Feature', orientation='h',
    title=f'Feature Importance — {best_model_name}',
    text=fi_df['Importance'].round(3),
    color='Importance', color_continuous_scale=[[0,'#2a2a2a'],[1,'#c8a96e']],
)
fig.update_traces(textposition='outside', textfont=dict(color='#f0ece4'))
fig.update_layout(**DARK, xaxis_title='Importance', yaxis_title='',
                  showlegend=False, coloraxis_showscale=False)
fig.write_html('public/charts/whr_feature_importance.html', full_html=True, include_plotlyjs='cdn')
fig.show()


### 2026 Predictions

Using 2025 factor values as a proxy for 2026 inputs, the best model generates projected happiness scores and rankings for the coming year. These are model projections under a ceteris paribus assumption — no geopolitical shocks, no structural changes. Treat them as a baseline, not a forecast.

In [29]:
# 2026 predictions — using best model retrained on all available data
X_2026 = master_df[master_df['Year'] == 2025][features]
y_2026_predicted = best_model.predict(X_2026)

predicted_2026_df = pd.DataFrame({
    'Country'                   : master_df[master_df['Year'] == 2025]['Country'].values,
    'Predicted_Score_2026'      : y_2026_predicted,
})
predicted_2026_df['Predicted_Rank_2026'] = predicted_2026_df['Predicted_Score_2026'].rank(ascending=False)

actual_2025_df = master_df[master_df['Year'] == 2025][['Country','Happiness_Score']].copy()
actual_2025_df['Actual_Rank_2025'] = actual_2025_df['Happiness_Score'].rank(ascending=False)

rank_comparison_df = pd.merge(predicted_2026_df, actual_2025_df, on='Country')
rank_comparison_df['Rank_Change'] = rank_comparison_df['Actual_Rank_2025'] - rank_comparison_df['Predicted_Rank_2026']

print("Predicted Top 10 — 2026:")
print(rank_comparison_df.sort_values('Predicted_Rank_2026').head(10)[
    ['Country','Predicted_Score_2026','Predicted_Rank_2026','Rank_Change']
].to_string(index=False))


Predicted Top 10 — 2026:
    Country  Predicted_Score_2026  Predicted_Rank_2026  Rank_Change
    Denmark              7.507456                  1.0          2.0
New Zealand              7.498027                  2.0          9.0
     Norway              7.475512                  3.0          3.0
     Sweden              7.398424                  4.0          1.0
  Singapore              7.393860                  5.0         31.0
    Ireland              7.376654                  6.0          7.0
    Germany              7.366549                  7.0         10.0
Netherlands              7.365718                  8.0         -1.0
    Finland              7.354502                  9.0         -8.0
Switzerland              7.347016                 10.0          0.0


### Canada Deep Dive

Canada has experienced one of the most significant ranking declines among developed nations — dropping from the top 10 in 2015 to outside the top 15 by 2025. The factor profile reveals the likely culprits: deteriorating scores on **social support and freedom**, compounded by cost-of-living pressures reflected in the GDP contribution gap versus Nordic peers.

The model projects a modest recovery in 2026 — but a return to Canada's 2015 standing would require structural improvements, not just cyclical reversion.

In [30]:
# Canada deep dive — historical trend + 2026 prediction
canada_df = master_df[master_df['Country'] == 'Canada'].copy()

fig = px.line(
    canada_df, x='Year', y='Happiness_Score',
    title='Canada — Happiness Score Trend (2015–2025)',
    markers=True,
)
fig.update_traces(line=dict(color=ACCENT, width=2.5), marker=dict(color=ACCENT, size=7))
fig.update_layout(**DARK, xaxis_title='Year', yaxis_title='Happiness Score',
                  xaxis=dict(**AXIS, tickmode='linear', dtick=1))
fig.show()

canada_2025_rank  = rank_comparison_df[rank_comparison_df['Country'] == 'Canada']['Actual_Rank_2025'].values[0]
canada_2026_score = rank_comparison_df[rank_comparison_df['Country'] == 'Canada']['Predicted_Score_2026'].values[0]
canada_2026_rank  = rank_comparison_df[rank_comparison_df['Country'] == 'Canada']['Predicted_Rank_2026'].values[0]

print(f"2025 Rank:             #{int(canada_2025_rank)}")
print(f"Predicted 2026 Score:  {canada_2026_score:.3f}")
print(f"Predicted 2026 Rank:   #{int(canada_2026_rank)}")

# Canada factor profile
canada_factors = canada_df[canada_df['Year'] == 2025][features].melt(var_name='Factor', value_name='Value')
fig2 = px.bar(
    canada_factors, x='Factor', y='Value',
    title="Canada — Happiness Factor Profile (2025)",
    text=canada_factors['Value'].round(3),
    color='Value', color_continuous_scale=[[0,'#2a2a2a'],[1,'#c8a96e']],
)
fig2.update_traces(textposition='outside', textfont=dict(color='#f0ece4'))
fig2.update_layout(**DARK, xaxis_title='', yaxis_title='Contribution to Score',
                   showlegend=False, coloraxis_showscale=False)
fig2.show()


2025 Rank:             #25
Predicted 2026 Score:  7.313
Predicted 2026 Rank:   #11


---
## Key Takeaways

1. **Global happiness is more resilient than expected.** COVID did not produce the collapse the data seemed to promise — post-pandemic scores are higher than pre-pandemic averages.
2. **Three factors dominate:** GDP per capita, social support, and life expectancy account for the bulk of cross-country happiness variation. Corruption is the primary negative driver.
3. **The G7 paradox.** The world's wealthiest bloc ranks outside the top 10 — consistently outperformed by smaller Nordic nations with stronger social infrastructure.
4. **Canada is in structural decline.** A drop from top-10 to outside top-15 over a decade is not noise — it reflects real deterioration in the factors that matter.
5. **COVID's legacy is country-specific.** The aggregate looks stable, but Afghanistan and Lebanon show what extreme political and economic collapse does to national wellbeing.
6. **The best model (Random Forest) achieves strong predictive performance**, with GDP per capita and social support as its most important features — consistent with the correlation analysis.